contrib_bots/lib/gcalendar.py

```diff
@@ -0,0 +1,233 @@
+            to=original_stream,
+            subject=original_subject,
+            content=msg,
+        ))
+    else:
+        raise 'Unknown message type'
+
+class GCalendarHandler(object):
+    '''
+    This plugin facilitates creating Google Calendar events.
+    '''
+    def usage(self):
+        return '''
+            This plugin will allow users to create events
+            in a Google Calendar. Users should preface
+            messages with "@gcalendar"
+
+            Before running this, make sure to register a new
+            project in the Google Developers Console
+            (https://console.developers.google.com/start/api?id=calendar)
+            and download the "client secret" to ~/.gcal_client_secret.json
+            '''
+
+    def triage_message(self, message):
+        original_content = message['content']
+
+        is_gcalendar = original_content.startswith('@gcalendar')
+
+        return is_gcalendar
+
+    def handle_message(self, message, client, state_handler):
+        original_content = message['content']
+        original_sender = message['sender_email']
+
+        content = message['content'].replace('@gcalendar ', '')
+
+        # Identify event name and date (ISO 8601 notation)
+        #   Expected formats:
+        #       "Meeting with John" 2003-01-14T13:37:00+01:00 2003-01-14T15:37:00+01:00
+        #       Trip 2003-01-14
+        input_re = '(\w+|"[\S ]+") ((\d{4}-(0[1-9]|1[1-2])-(0[1-9]|1[1-9]|2[1-9]|3[0-1])T(0[0-9]|1[0-9]|2[0-3]):[0-5][0-9]:[0-5][0-9](\+|\-)(0[1-9]|1[1-4])\:[03]0 ?){2}|(\d{4}-(0[1-9]|1[1-2])-(0[1-9]|1[1-9]|2[1-9]|3[0-1])$))'
+
+        m = re.match(input_re, content)
+
+        if not m:
+            error = ('Unknown arguments set.\n'
+                'Usage examples:\n\n'
+                '```\n'
+                '@gcalendar "Meeting with John" 2003-01-14T13:37:00+01:00 2003-01-14T15:37:00+01:00\n'
+                '@gcalendar Trip 2003-01-14\n'
+                '```\n\n'
+                'Remember to use the [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) notation for dates: '
+                '`2020-04-16T13:58:13-02:00`')
+
+            reply_message(error, message, client)
+
+            return
+
+        title = m.group(1).strip('"')
+
+        splits = m.group(2).split();
+        start_str = splits[0]
+
+        event = {
+            'summary': title
+        }
+
+        if len(splits) != 2: # Full-day event
+            end_str = None
+            event.update({
+                'start': {
+                    'date': start_str
+                },
+                'end': {
+                    'date': start_str
+                }
+            })
+        else:
+            end_str = splits[1]
+            event.update({
+                'start': {
+                    'dateTime': start_str
+                },
+                'end': {
+                    'dateTime': end_str
+                }
+            })
+
+        credentials = get_credentials()
```

Comment:
The thing is that I use `credentials` [here](https://github.com/zulip/zulip/pull/2771/files/8ab6ab503588b3cc80ad03aba2d6d56ec06c7ca0#diff-94006fd9e4c82133827aef81c1463418R81), too. This means I cannot use `__init__`, because `list_calendars()` is outside the class.

What about maybe using a global variable?